sdk/python/lib/pulumi/runtime/stack.py

```diff
@@ -50,28 +59,32 @@ async def run_pulumi_func(func: Callable):
             log.debug(f"waiting for quiescence; {rpcs_remaining} RPCs outstanding")
             await RPC_MANAGER.rpcs.pop()
 
-        # Asyncio event loops require that all outstanding tasks be completed by the time that the
-        # event loop closes. If we're at this point and there are no outstanding RPCs, we should
-        # just cancel all outstanding tasks.
-        #
-        # We will occasionally start tasks deliberately that we know will never complete. We must
-        # cancel them before shutting down the event loop.
-        log.debug("Canceling all outstanding tasks")
-        for task in _all_tasks():
-            # Don't kill ourselves, that would be silly.
-            if task == _get_current_task():
-                continue
-            task.cancel()
-
-        # Pump the event loop again. Task.cancel is delivered asynchronously to all running tasks
-        # and each task needs to get scheduled in order to acknowledge the cancel and exit.
-        await asyncio.sleep(0)
-
-        # Once we get scheduled again, all tasks have exited and we're good to go.
-        log.debug("run_pulumi_func completed")
+        if RPC_MANAGER.unhandled_exception is not None:
+            raise RPC_MANAGER.unhandled_exception.with_traceback(RPC_MANAGER.exception_traceback)
+
+        log.debug("RPCs successfully completed")
 
-    if RPC_MANAGER.unhandled_exception is not None:
-        raise RPC_MANAGER.unhandled_exception.with_traceback(RPC_MANAGER.exception_traceback)
+        # If the RPCs have successfully completed, now await all remaining outstanding tasks.
+        outstanding_tasks = _get_running_tasks()
+        if len(outstanding_tasks) == 0:
+            log.debug("No outstanding tasks to complete, exiting")
+        else:
+            log.debug(f"Waiting for {len(outstanding_tasks)} outstanding tasks to complete")
+
+            done, pending = await asyncio.wait(outstanding_tasks, return_when="FIRST_EXCEPTION")
```

Comment:
It doesn't look like `done` is used elsewhere. Not sure if it's needed?

Reply:
It is now. The latest commit uses the done pile to find and raise exceptions.